rpcs3/Emu/Cell/lv2/sys_uart.cpp

```diff
@@ -1,35 +1,597 @@
+			}
+
+			vuart_thread.rx_wake_c.wait_unlock(5000, lock);
+		}
+
+		ppu.check_state();
+	}
+	else
+	{
+		const s32 read_result = vuart_read(data.get(), static_cast<u32>(size));
+
+		if (read_result <= CELL_OK)
+		{
+			return read_result;
+		}
+
+		read_size = read_result;
+	}
+
+	if (!buffer || !vm::check_addr(buffer.addr(), vm::page_writable, read_size))
+	{
+		return CELL_EFAULT;
+	}
+
+	memcpy(buffer.get_ptr(), data.get(), read_size);
+	return not_an_error(read_size);
 }
 
-error_code sys_uart_send(vm::cptr<void> buffer, u64 size, u64 flags)
+error_code sys_uart_send(ppu_thread &ppu, vm::cptr<void> buffer, u64 size, u32 mode)
 {
-	sys_uart.todo("sys_uart_send(buffer=0x%x, size=0x%llx, flags=0x%x)", buffer, size, flags);
+	sys_uart.trace("sys_uart_send(buffer=0x%x, size=0x%llx, mode=0x%x)", buffer, size, mode);
 
-	return CELL_OK;
+	if (!size)
+	{
+		return CELL_OK;
+	}
+
+	if ((mode & ~3ul) != 0)
+	{
+		return CELL_EINVAL;
+	}
+
+	auto &vuart_thread = g_fxo->get<vuart_av>();
+
+	if (!vuart_thread.init)
+	{
+		return CELL_ESRCH;
+	}
+
+	if (size > 0x20000U)
+	{
+		// kmalloc restriction
+		fmt::throw_exception("Buffer is too big");
+	}
+
+	if (!vm::check_addr(buffer.addr(), vm::page_readable, static_cast<u32>(size)))
+	{
+		return CELL_EFAULT;
+	}
+
+	const std::unique_ptr<u8[]> data = std::make_unique<u8[]>(size);
+	memcpy(data.get(), buffer.get_ptr(), size);
+
+	std::unique_lock lock(vuart_thread.tx_mutex, std::defer_lock);
+
+	const u32 ITER_SIZE = 4096;
+
+	if (mode & 1)
+	{
+		// Yield before sending packets
+		lv2_obj::sleep(ppu);
+
+		lock.lock();
+
+		auto vuart_send_all = [&](const u8 *data, u32 data_sz)
+		{
+			u32 rem_size = data_sz;
+
+			while (rem_size)
+			{
+				if (ppu.is_stopped())
+				{
+					return false;
+				}
+
+				std::unique_lock<shared_mutex> lock(vuart_thread.tx_rdy_m);
+				if (vuart_thread.get_tx_bytes() >= PS3AV_TX_BUF_SIZE)
+				{
+					vuart_thread.tx_rdy_c.wait_unlock(5000, lock);
+				}
+				else
+				{
+					lock.unlock();
+				}
+				rem_size -= vuart_thread.enque_tx_data(data + data_sz - rem_size, rem_size);
+			}
+
+			return true;
+		};
+
+		if (!vuart_send_all(data.get(), std::min(ITER_SIZE, static_cast<u32>(size)))) return {};
+
+		if (size > ITER_SIZE)
+		{
+			u32 sent_size;
+
+			for (sent_size = ITER_SIZE; sent_size + ITER_SIZE < size; sent_size += ITER_SIZE)
+			{
+				if (!vuart_send_all(data.get() + sent_size, ITER_SIZE)) return {};
+			}
+
+			if (!vuart_send_all(data.get() + sent_size, static_cast<u32>(size) - sent_size)) return {};
+		}
+
+		ppu.check_state();
+	}
+	else if (mode & 2)
+	{
+		if (!lock.try_lock())
+		{
+			return CELL_EBUSY;
+		}
+
+		if (PS3AV_TX_BUF_SIZE - vuart_thread.get_tx_bytes() < size)
+		{
+			return CELL_EAGAIN;
+		}
+
+		return not_an_error(vuart_thread.enque_tx_data(data.get(), static_cast<u32>(size)));
+	}
+	else
+	{
+		if (!lock.try_lock())
+		{
+			return CELL_EBUSY;
+		}
+
+		const u32 first_iter_sz   = std::min(ITER_SIZE, static_cast<u32>(size));
```

Comment:
Ditto.

Reply:
This one has a bit weird return values based on first, middle and last iteration

Reply:
I don't see this bit anymore. Is it fixed and working?

Reply:
Lines 346-373

Reply:
Gotcha. Looks good.